Update step numbers in URL constructor

ohno418 opened the pull request:
<!-- Please describe your changes on the following line: -->
Two commits have been made ahead of the current implementation of
Servo's URL constructor:

- Align with a more modern IDL definition style [1]
- Add URL.canParse() [2]

Since these commits don't alter the actual behavior, this commit only
updates the step numbers and adds brief descriptions for each step.

No behavior change is expected with this commit.

[1]: https://github.com/whatwg/url/commit/ea3b75d333cbce91f73676ef2d21bcadb12b4419
[2]: https://github.com/whatwg/url/commit/ae3c28b84e3e7122c2807401c26b8a63cb2ab445


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [x] These changes do not require tests because no behavior has changed.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

ohno418 commented on components/script/dom/url.rs:
```diff
@@ -77,38 +77,34 @@ impl URL {
-        // Step 5: Skip (see step 8 below).
-        // Steps 6-7.
-        let result = URL::new(global, parsed_url);
-        // Step 8: Instead of construcing a new `URLSearchParams` object here, construct it
-        //         on-demand inside `URL::SearchParams`.
```
I don't know exactly why this happened, but this PR doesn't change the behavior around `URLSearchParams`, continuing to ignore the related steps.

mrobinson replied:
Is this change just updating the step numbers?

ohno418 replied:
Yes. The corresponding spec change related to this PR can be found here: https://github.com/whatwg/url/commit/ae3c28b8

(I probably should have mentioned this somewhere, such as in commit message.)

mrobinson replied:
I'm not sure the code changes are an improvement here, because they increase the nesting level of the method. Would it be possible to avoid changing the code if there is no behavior changes and also to paste a small description of each step in quotes. With a small amount of text, at least it will still be clear what each step does.

ohno418 replied:
Okay, that makes sense. I'll make the change.

ohno418 committed:
Update step numbers in URL constructor

Two commits have been made ahead of the current implementation of
Servo's URL constructor:

- Align with a more modern IDL definition style [1]
- Add URL.canParse() [2]

Since these commits don't alter the actual behavior, this commit only
updates the step numbers and adds brief descriptions for each step.

No behavior change is expected with this commit.

[1]: https://github.com/whatwg/url/commit/ea3b75d333cbce91f73676ef2d21bcadb12b4419
[2]: https://github.com/whatwg/url/commit/ae3c28b84e3e7122c2807401c26b8a63cb2ab445

Signed-off-by: Yutaro Ohno <yutaro.ono.418@gmail.com>

ohno418 commented:
As I mentioned in the commit message, the current implementation status of the URL constructor appears to correspond to the state before the following commit of the spec:

https://github.com/whatwg/url/commit/ea3b75d333cbce91f73676ef2d21bcadb12b4419

ohno418 commented on components/script/dom/url.rs:
```diff
@@ -77,38 +77,38 @@ impl URL {
+        // Instead of construcing a new `URLSearchParams` object here, construct it
+        // on-demand inside `URL::SearchParams`.
+        //
+        // Step 3. Let query be parsedURL’s query.
+        // Step 5. Set this’s query object to a new URLSearchParams object.
+        // Step 6. Initialize this’s query object with query.
+        // Step 7. Set this’s query object’s URL object to this.
```
This is a bit confusing. Step 5 and 8, which were previously skipped, seem to correspond these four steps.

Step 8 is split into some parts here:
https://github.com/whatwg/url/commit/ea3b75d333cbce91f73676ef2d21bcadb12b4419#diff-29243b3b9b716b55c6a61970b0c4864f464b139d397fb961a05bb6e1e2b97cabL2840-L2842

Here's the current spec:
https://url.spec.whatwg.org/#constructors

mrobinson commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 3336bf6 has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson 3336bf6cce3bc1caba1ff440e507097285b2eeb8 -->

bors-servo commented:
:hourglass: Testing commit 3336bf6cce3bc1caba1ff440e507097285b2eeb8 with merge c67d76026463c573c453d298e34a8bfde84d8702...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/4912618224/jobs/8774156505)
Approved by: mrobinson
Pushing c67d76026463c573c453d298e34a8bfde84d8702 to master...
